File: bilibili/src/main/java/com/bilibili/ui/bangumi/BangumiFragment.java
package com.bilibili.ui.bangumi;

import android.os.SystemClock;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.MotionEvent;

import com.bilibili.App;
import com.bilibili.R;
import com.bilibili.model.bean.bangumi.BangumiIndexFall;
import com.bilibili.model.bean.bangumi.BangumiIndexPage;
import com.bilibili.ui.bangumi.viewbinder.BangumiDividerBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiHomeBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexFallBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexFollowBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexPageFootBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexRecommendBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiRecommendDetailBinder;
import com.bilibili.widget.recyclerview.CommonAdapter;
import com.common.base.BaseMvpFragment;
import com.common.widget.adapter.base.BaseAdapterWrapper;

import butterknife.BindView;
import me.drakeet.multitype.Items;

/**
 * Created by miserydx on 17/6/29.
 */

public class BangumiFragment extends BaseMvpFragment<BangumiPresenter> implements BangumiContract.View {

    public static final String TAG = BangumiFragment.class.getSimpleName();

    private static int SPAN_COUNT = 3;

    @BindView(R.id.rv)
    RecyclerView mRecyclerView;
    @BindView(R.id.layout_refresh)
    SwipeRefreshLayout mRefreshLayout;

    private CommonAdapter mAdapter;

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_bangumi;
    }

    @Override
    protected void initInject() {
        App.getInstance().getFragmentComponent().inject(this);
    }

    @Override
    protected void initViewAndEvent() {
        mRefreshLayout.setColorSchemeResources(R.color.theme_color_primary);
        mRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                if (mAdapter.getItems().size() == 0) {
                    mPresenter.loadData();
                } else {
                    mPresenter.pullToRefresh();
                }
            }
        });
        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), SPAN_COUNT);
        GridLayoutManager.SpanSizeLookup spanSizeLookup = new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                Object o = mAdapter.getItems().get(position);
                return o instanceof BangumiIndexPage.Recommend ? 1 : SPAN_COUNT;
            }
        };
        layoutManager.setSpanSizeLookup(spanSizeLookup);
        mRecyclerView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.white));
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.addItemDecoration(new BangumiIndexItemDecoration());
        mAdapter = new CommonAdapter();
        //register item
        mAdapter.register(BangumiIndexFollowBinder.BangumiIndexFollow.class, new BangumiIndexFollowBinder());
        mAdapter.register(BangumiHomeBinder.BangumiHome.class, new BangumiHomeBinder());
        mAdapter.register(BangumiIndexRecommendBinder.BangumiIndexRecommend.class, new BangumiIndexRecommendBinder());
        mAdapter.register(BangumiIndexPage.Recommend.class, new BangumiRecommendDetailBinder());
        mAdapter.register(BangumiIndexPage.Foot.class, new BangumiIndexPageFootBinder());
        mAdapter.register(BangumiIndexFall.class, new BangumiIndexFallBinder());
        mAdapter.register(BangumiDividerBinder.BangumiDivider.class, new BangumiDividerBinder());
        //other settings
        mAdapter.setOnLoadMoreListener(new CommonAdapter.OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                mPresenter.loadMore();
            }
        });
        mAdapter.setOnClickRetryListener(new BaseAdapterWrapper.OnClickRetryListener() {
            @Override
            public void onClickRetry() {
                mPresenter.loadMore();
            }
        });
        mAdapter.setScrollSaveStrategyEnabled(true);
        mRecyclerView.setAdapter(mAdapter);
    }

    @Override
    public void onDataUpdated(Items items, int state) {
        switch (state) {
            case BangumiPresenter.STATE_INITIAL:
                mAdapter.setItems(items);
                mAdapter.notifyDataSetChanged();
                break;
            case BangumiPresenter.STATE_REFRESHING:
                Items temp = new Items();
                temp.addAll(mAdapter.getItems().subList(items.size(), mAdapter.getItems().size()));
                items.addAll(temp);
                mAdapter.setItems(items);
                mAdapter.notifyDataSetChanged();
                break;
            case BangumiPresenter.STATE_LOAD_MORE:
                if (items.size() == 0) {
                    mAdapter.showNoMore();
                    return;
                }
                mAdapter.addItems(items);
                mRecyclerView.dispatchTouchEvent(MotionEvent.obtain(SystemClock.uptimeMillis(), SystemClock.uptimeMillis(), MotionEvent.ACTION_CANCEL, 0f, 0f, 0));//停止recyclerView滑动
                mAdapter.loadMoreComplete();
                break;
        }
    }

    @Override
    public void onRefreshingStateChanged(boolean isRefresh) {
        mRefreshLayout.setRefreshing(isRefresh);
    }

    @Override
    public void showLoadMoreError() {
        mAdapter.showFailToLoadMore();
    }

    @Override
    public void showLoadFailed() {
        mAdapter.showLoadFailed();
    }
}

File: bilibili/src/main/java/com/bilibili/ui/bangumi/BangumiFragment.kt
package com.bilibili.ui.bangumi

import android.support.v4.content.ContextCompat
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView

import com.bilibili.App
import com.bilibili.R
import com.bilibili.widget.recyclerview.BiliMultiTypeAdapter
import com.common.base.BaseMvpFragment

import butterknife.BindView
import com.bilibili.model.bean.BangumiIndexFall
import com.bilibili.model.bean.Foot
import com.bilibili.model.bean.Recommend
import com.bilibili.ui.bangumi.viewbinder.*
import me.drakeet.multitype.Items

/**
 * Created by miserydx on 17/6/29.
 */

class BangumiFragment : BaseMvpFragment<BangumiPresenter>(), BangumiContract.View {

    companion object {
        val TAG = BangumiFragment::class.java.simpleName

        private val SPAN_COUNT = 3
    }

    @BindView(R.id.rv)
    lateinit var mRecyclerView: RecyclerView
    @BindView(R.id.layout_refresh)
    lateinit var mRefreshLayout: SwipeRefreshLayout

    private lateinit var mAdapter: BiliMultiTypeAdapter
    private var items = Items()

    override fun getLayoutId(): Int = R.layout.fragment_bangumi

    override fun initInject() {
        App.getInstance().fragmentComponent.inject(this)
    }

    override fun initViewAndEvent() {
        mRefreshLayout.setColorSchemeResources(R.color.theme_color_primary)
        mRefreshLayout.setOnRefreshListener { mPresenter.pullToRefresh() }
        val layoutManager = GridLayoutManager(context, SPAN_COUNT)
        val spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val o = items[position]
                return if (o is Recommend) 1 else SPAN_COUNT
            }
        }
        layoutManager.spanSizeLookup = spanSizeLookup
        mRecyclerView.setBackgroundColor(ContextCompat.getColor(context!!, R.color.white))
        mRecyclerView.layoutManager = layoutManager
        mRecyclerView.addItemDecoration(BangumiIndexItemDecoration(spanSizeLookup))
        mAdapter = BiliMultiTypeAdapter()
        //register item
        mAdapter.register(BangumiIndexFollowBinder.BangumiIndexFollow::class.java, BangumiIndexFollowBinder())
        mAdapter.register(BangumiIndexFollowBinder.BangumiIndexFollow::class.java, BangumiIndexFollowBinder())
        mAdapter.register(BangumiHomeBinder.BangumiHome::class.java, BangumiHomeBinder())
        mAdapter.register(BangumiIndexRecommendBinder.BangumiIndexRecommend::class.java, BangumiIndexRecommendBinder())
        mAdapter.register(Recommend::class.java, BangumiRecommendDetailBinder())
        mAdapter.register(Foot::class.java, BangumiIndexPageFootBinder())
        mAdapter.register(BangumiIndexFall::class.java, BangumiIndexFallBinder())
        mAdapter.register(BangumiDividerBinder.BangumiDivider::class.java, BangumiDividerBinder())
        mAdapter.setFooterItemEnabled(true)
        mAdapter.setOnLoadMoreListener({ mPresenter.loadMore() }, BiliMultiTypeAdapter.LOAD_MORE_MODE_BOTTOM)
        mAdapter.setScrollSaveStrategyEnabled(true)
        mRecyclerView.adapter = mAdapter
    }

    @Synchronized override fun onDataUpdated(items: Items?, state: Int) {
        when (state) {
            BangumiPresenter.STATE_INITIAL -> {
                this.items = items!!
                mAdapter.items = this.items
                mAdapter.notifyDataSetChanged()
            }
            BangumiPresenter.STATE_REFRESHING -> {
                val temp = Items()
                temp.addAll(this.items.subList(items!!.size, this.items.size - 1))
                this.items = items
                this.items.addAll(temp)
                mAdapter.items = this.items
                mAdapter.notifyDataSetChanged()
            }
            BangumiPresenter.STATE_LOAD_MORE -> {
                if (items!!.size == 0) {
                    mAdapter.showNoMore()
                }
                val position = this.items.size
                this.items.addAll(items)
                mAdapter.items = this.items
                mAdapter.setLoadMoreFinished()
                mAdapter.notifyItemInserted(position)
            }
            BangumiPresenter.STATE_LOAD_ERROR -> mAdapter.showFailToLoad()
        }

    }

    override fun onRefreshingStateChanged(isRefresh: Boolean) {
        mRefreshLayout.isRefreshing = isRefresh
    }
}
